juriscraper/pacer/appellate_attachment_page.py

```diff
@@ -0,0 +1,212 @@
+
+
+class AppellateAttachmentPage(BaseReport):
+    """An object for querying and parsing the appellate att. page report."""
+
+    PATH = "doc1/"
+
+    def __init__(self, court_id, pacer_session=None):
+        super(AppellateAttachmentPage, self).__init__(court_id, pacer_session)
+        # Note that parsing bankruptcy attachment pages does not reveal the
+        # document number, only the attachment numbers.
+        self.is_bankruptcy = self.court_id.endswith("b")
+
+    def query(self, document_number):
+        """Query the "attachment page" endpoint and set the results to self.response.
+
+        :param document_number: The internal PACER document ID for the item.
+        :return: a request response object
+        """
+        assert (
+            self.session is not None
+        ), "session attribute of DocketReport cannot be None."
+        # coerce the fourth digit of the document number to 1 to ensure we get
+        # the attachment page.
+        document_number = document_number[:3] + "0" + document_number[4:]
+        url = self.url + document_number
+        logger.info(u"Querying the attachment page endpoint at URL: %s", url)
+        self.response = self.session.get(url)
+        self.parse()
+
+    def _strip_bad_html_tags_insecure(self, tree):
+        """Remove bad tags from HTML while leaving scripts.
+
+        :param tree: A tree you wish to cleanup
+        :type tree: lxml.html.HtmlElement
+        Set self.tree to a cleaned lxml.html.HtmlElement
+        :return: None
+        """
+        self.tree = strip_bad_html_tags_insecure(tree, remove_scripts=False)
+
+    @property
+    def data(self):
+        """Get data back from the query for the matching document entry.
+
+        Appellate attachments is different than the district court.  We have
+        no document_id. Instead we can use the pacer_seq_no or the dls_id
+        to discern which attachment goes with which docket entry.
+
+        Additionally, the main document is not identifiable in attachments.
+        It is presumed to be the document with the lowest pacer_doc_id.
+
+        :return: If lookup fails, an empty dict. Else, a dict containing the
+        following fields:
+            - dls_id: The id used to generate the POST request for this page.
+            - case_id: The pacer case id for this case.
+            - pacer_seq_no - The value of the docket row we came from.
+
+            - attachments: A list of attached items with the following fields:
+                - attachment_number: The attachment number.
+                - description: A description of the item.
+                - page_count: The number of pages for the attachment.
+                - pacer_doc_id: The document ID for the attachment (a str).
+
+        See the JSON objects in the tests for more examples.
+        """
+        if self.is_valid is False:
+            return {}
+
+        rows = self.tree.xpath("//tbody/tr/td/a/parent::td/parent::tr")
```

Comment:
As a matter of style, I try to have comments showing the HTML structure that we're parsing, especially if the XPath at all feels challenging to eyeball. This is probably too much to be asking for?

Reply:
√ Sounds good.